### PR TITLE
fix(curriculum) - added new tests to prevent invalid solutions in the Remove Elements from a Linked List by Index challenge

### PR DESCRIPTION
**Pull request by RandellDawson**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

While reviewing PR #35825, I discovered some new tests were needed to make sure only valid solutions would pass the challenge.  

**Note:** I removed the `remove` method from this challenge as noted on PR #35825 that it is not used in the tests and could be deleted.

### CODE REVIEW

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 1]
```diff
@@ -25,15 +24,19 @@ Remember to keep count of the <code>currentIndex</code>.
...
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); var removedItem = test.removeAt(0); return test.head() === null && test.size() === 0 && removedItem === 'cat';}()));        
+  - text: Your <code>removeAt</code> method should return the element of the removed node.
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten');  return test.removeAt(1) === 'dog'}()));
+  - text: Your <code>removeAt</code> method should return <code>null</code> and the list not change if the given index is less than <code>0</code>
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten');  var removedItem = test.removeAt(-1); return removedItem === null && JSON.stringify(test.head()) === '{"element":"cat","next":{"element":"dog","next":{"element":"kitten","next":null}}}'}()));
+  - text: Your <code>removeAt</code> method should return <code>null</code> and the list should not change if the given index is equal or more than the <code>length</code> of the linked list.
```
Comment (Manish-Giri) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
  - text: Your <code>removeAt</code> method should return <code>null</code> and the linked list should not change if the given index is equal to or more than the <code>length</code> of the list.
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 2]
```diff
@@ -25,15 +24,19 @@ Remember to keep count of the <code>currentIndex</code>.
...
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten'); test.add('bird');test.removeAt(1); return JSON.stringify(test.head()) === '{"element":"cat","next":{"element":"kitten","next":{"element":"bird","next":null}}}'}())); 
+  - text: When only one element is present in the linked list, your <code>removeAt</code> method should return the element at specified index, remove the element, and reduce the length of the linked list.
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); var removedItem = test.removeAt(0); return test.head() === null && test.size() === 0 && removedItem === 'cat';}()));        
+  - text: Your <code>removeAt</code> method should return the element of the removed node.
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten');  return test.removeAt(1) === 'dog'}()));
+  - text: Your <code>removeAt</code> method should return <code>null</code> and the list not change if the given index is less than <code>0</code>
```
Comment (Manish-Giri) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
  - text: Your <code>removeAt</code> method should return <code>null</code> and the linked list should not change if the given index is less than <code>0</code>.
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 3]
```diff
@@ -25,15 +24,19 @@ Remember to keep count of the <code>currentIndex</code>.
...
+    testString: assert((function(){var test = new LinkedList(); return (typeof test.removeAt === 'function')}()));
+  - text: Your <code>removeAt</code> method should reduce the <code>length</code> of the linked list.
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten'); test.removeAt(1); return test.size() === 2}()));
+  - text: Your <code>removeAt</code> method should remove the element at the specified index from the linked list.
+    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten'); test.add('bird');test.removeAt(1); return JSON.stringify(test.head()) === '{"element":"cat","next":{"element":"kitten","next":{"element":"bird","next":null}}}'}())); 
+  - text: When only one element is present in the linked list, your <code>removeAt</code> method should return the element at specified index, remove the element, and reduce the length of the linked list.
```
Comment (Manish-Giri) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
  - text: When only one element is present in the linked list, your <code>removeAt</code> method should remove and return the element at specified index, and reduce the length of the linked list.
```

Might be a good idea to simplify things.

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md (hunk 1 above)
Comment (RandellDawson) [reply curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]: @Manish-Giri What if I changed the wording to the following?
```
  - text: Your <code>removeAt</code> method should return <code>null</code> and the linked list should not change if the given index is greater than or equal to the <code>length</code> of the list.
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md (hunk 3 above)
Comment (RandellDawson) [reply curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]: Should I update the testString too?

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 4]
```diff
@@ -25,15 +24,19 @@ Remember to keep count of the <code>currentIndex</code>.
...
-  - text: Your <code>removeAt</code> method should also return <code>null</code> if the given index is less than <code>0</code>
-    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten');  return (test.removeAt(-1) === null)}()), 'Your <code>removeAt</code> method should also return <code>null</code> if the given index is less than <code>0</code>');
-  - text: Your <code>removeAt</code> method should also return <code>null</code> if the given index is equal or more than the <code>length</code> of the linked list.
-    testString: assert((function(){var test = new LinkedList(); test.add('cat'); test.add('dog'); test.add('kitten');  return (test.removeAt(3) === null)}()), 'Your <code>removeAt</code> method should also return <code>null</code> if the given index is equal or more than the <code>length</code> of the linked list.');
+    testString: assert((function(){var test = new LinkedList(); return (typeof test.removeAt === 'function')}()));
+  - text: Your <code>removeAt</code> method should reduce the <code>length</code> of the linked list.
```
Comment (Manish-Giri) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
  - text: Your <code>removeAt</code> method should reduce the <code>length</code> of the linked list by one.
```
I would prefer if we were direct, but I'll leave this up for other opinions.

**Review by Manish-Giri — CHANGES_REQUESTED** [review]
Some minor suggestions for the assertion messages.

Also, might be a good idea to decrement the `length` and return `removed` in just one place -
```js
    var removed = null;
    if (index >= length || index < 0) {
      return null;
    }
    if (index === 0) {
      removed = head.element;
      head = currentNode.next;
    } else {
      while (count < index) {
        previous = currentNode;
        currentNode = currentNode.next;
        count++;
      }
      removed = previous.next.element;
      previous.next = currentNode.next;
    }
    length--;
    return removed;
  };
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md (hunk 1 above)
Comment (Manish-Giri) [reply curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]: @RandellDawson Much cleaner!

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md (hunk 3 above)
Comment (Manish-Giri) [reply curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]: Which part?

The corresponding `testString` looks fine to me (unless I missed something) -
```js
 testString: assert((function(){var test = new LinkedList(); test.add('cat'); var removedItem = test.removeAt(0); return test.head() === null && test.size() === 0 && removedItem === 'cat';}())); 
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md (hunk 1 above)
Comment (RandellDawson) [reply curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]: Change your suggestion to mine and I will commit it.

**Comment by RandellDawson** [comment]
@Manish-Giri I think I have addressed everything else you wanted in my latest commit.

**Review by Manish-Giri — APPROVED** [review]
@RandellDawson LGTM

**Review by ojeytonwilliams — APPROVED** [review]
LGTM.  There will be some merge conflicts with #35825 so I won't merge this just yet.

**Review by Manish-Giri — APPROVED** [review]
@RandellDawson LGTM. Thank You for your contribution.

**Pull request merged**